stl/inc/ranges

```diff
@@ -7049,6 +7049,493 @@ namespace ranges {
+                return _Sentinel<false>{_Tuple_transform(_RANGES end, _Views)};
+            } else if constexpr ((random_access_range<_ViewTypes> && ...)) {
+                return begin() + static_cast<iter_difference_t<_Iterator<false>>>(size());
+            } else {
+                return _Iterator<false>{_Tuple_transform(_RANGES end, _Views)};
+            }
+        }
+
+        _NODISCARD constexpr auto end() const noexcept(_Is_end_noexcept<true>) // strengthened
+            requires (range<const _ViewTypes> && ...)
+        {
+            if constexpr (!_Zip_is_common<const _ViewTypes...>) {
+                return _Sentinel<true>{_Tuple_transform(_RANGES end, _Views)};
+            } else if constexpr ((random_access_range<const _ViewTypes> && ...)) {
+                return begin() + static_cast<iter_difference_t<_Iterator<true>>>(size());
+            } else {
+                return _Iterator<true>{_Tuple_transform(_RANGES end, _Views)};
+            }
+        }
+
+        _NODISCARD constexpr auto size() noexcept(
+            noexcept(_STD apply(_Size_closure(), _Tuple_transform(_RANGES size, _Views)))) // strengthened
+            requires (sized_range<_ViewTypes> && ...)
+        {
+            return _STD apply(_Size_closure(), _Tuple_transform(_RANGES size, _Views));
+        }
+
+        _NODISCARD constexpr auto size() const
+            noexcept(noexcept(_STD apply(_Size_closure(), _Tuple_transform(_RANGES size, _Views)))) // strengthened
+            requires (sized_range<const _ViewTypes> && ...)
+        {
+            return _STD apply(_Size_closure(), _Tuple_transform(_RANGES size, _Views));
+        }
+
+    private:
+        _NODISCARD static constexpr auto _Size_closure() noexcept {
+            return [](auto... _Sizes) {
+                using _Common_unsigned_type = _Make_unsigned_like_t<common_type_t<decltype(_Sizes)...>>;
+                return (_RANGES min)({static_cast<_Common_unsigned_type>(_Sizes)...});
+            };
+        }
+    };
+
+    template <class... _RangeTypes>
+    zip_view(_RangeTypes&&...) -> zip_view<views::all_t<_RangeTypes>...>;
+
+    template <class... _ViewTypes>
+    inline constexpr bool enable_borrowed_range<zip_view<_ViewTypes...>> = (enable_borrowed_range<_ViewTypes> && ...);
+
+    namespace views {
+        struct _Zip_fn {
+        private:
+            template <class... _Types>
+            static _CONSTEVAL bool _Is_invocation_noexcept() {
+                if constexpr (sizeof...(_Types) == 0) {
+                    // NOTE: views::empty<tuple<>> is nothrow copy-constructible.
+                    return true;
+                } else {
+                    return noexcept(zip_view<views::all_t<_Types>...>(_STD declval<_Types>()...));
+                }
+            }
+
+        public:
+            // clang-format off
+            template <class... _Types>
+            constexpr auto operator()(_Types&&... _Args) const

```

Comment:
`zip`'s `operator()` should be `_NODISCARD`, as it would be pointless to call and then immediately drop on the floor without doing anything. (The helper `_Is_invocation_noexcept()` could also be marked.)